From a4ce91a80d40ba4e029908188222ab61c4f80229 Mon Sep 17 00:00:00 2001
From: "Gary E. Miller" <gem@rellim.com>
Date: Thu, 3 May 2012 15:57:31 -0700
Subject: [PATCH] Fix a typo in a previous patch. Fix a serious bug in
 ntpshm_put() caused by using ship_to_ntp, which is packet
 synchronous, in the PPS thread context which is
 asynchronous to packets. I'm not totally content yet that
 things are aas they should be.

---
 libgpsd_core.c |    2 +-
 ntpshm.c       |    8 ++++++--
 2 files changed, 7 insertions(+), 3 deletions(-)

--- a/libgpsd_core.c
+++ b/libgpsd_core.c
@@ -1069,7 +1069,7 @@ gps_mask_t gpsd_poll(struct gps_device_t
 	 * devices output fix packets on a regular basis, even when unable
 	 * to derive a good fix. Such packets should set STATUS_NO_FIX.
 	 */
-	if ( 0 != session->gpsdata.set & LATLON_SET) {
+	if ( 0 != (session->gpsdata.set & LATLON_SET)) {
 	    if ( session->gpsdata.status > STATUS_NO_FIX) {
 		session->context->fixcnt++;
 		session->fixcnt++;
--- a/ntpshm.c
+++ b/ntpshm.c
@@ -761,8 +761,12 @@ static /*@null@*/ void *gpsd_ppsmonitor(
 		    cycle, duration,
 		    (unsigned long)tv.tv_sec, (unsigned long)tv.tv_usec);
 
-	/*@ +boolint @*/
-	if (session->ship_to_ntpd || session->context->pps_hook != NULL) {
+	/*  only listen to PPS after 4 consecutive fixes, otherwise time
+	 *  will be inaccurate.
+	 *  Not sure yet how to handle uBlox UBX_MODE_TMONLY
+	 *  Do not use ship_to_ntp here since it is synced to packets
+	 *  and this thread is asynchonous to packets */
+	if ( 3 < session->fixcnt ) {
 	    /*
 	     * The PPS pulse is normally a short pulse with a frequency of
 	     * 1 Hz, and the UTC second is defined by the front edge. But we
